Userland/Services/DeviceMapper/DeviceEventLoop.cpp

```diff
@@ -0,0 +1,234 @@
+    TRY(Core::System::endgrent());
+    return {};
+}
+
+ErrorOr<void> DeviceEventLoop::register_new_device(bool block_device, MajorNumber major_number, MinorNumber minor_number)
+{
+    auto* possible_match = device_node_family_to_match_type(block_device, major_number, minor_number);
+    if (!possible_match)
+        return {};
+    auto& match = *possible_match;
+    auto& device_node_family = *TRY(find_or_register_new_device_node_family(match, block_device, major_number, minor_number));
+
+    static constexpr StringView devtmpfs_base_path = "/dev/"sv;
+    auto path_pattern = TRY(String::from_utf8(match.path_pattern));
+    auto& allocation_map = device_node_family.devices_symbol_suffix_allocation_map();
+    auto possible_allocated_suffix_index = allocation_map.find_first_unset();
+    if (!possible_allocated_suffix_index.has_value()) {
+        // FIXME: Make the allocation map bigger?
+        return Error::from_errno(ERANGE);
+    }
+    auto allocated_suffix_index = possible_allocated_suffix_index.release_value();
+
+    auto path = TRY(String::from_utf8(path_pattern));
+    if (match.path_pattern.contains("%d"sv)) {
+        auto replacement = TRY(build_suffix_with_numbers(allocated_suffix_index));
+        path = TRY(path.replace("%d"sv, replacement, ReplaceMode::All));
+    }
+    if (match.path_pattern.contains("%c"sv)) {
+        auto replacement = TRY(build_suffix_with_letters(allocated_suffix_index));
+        path = TRY(path.replace("%c"sv, replacement, ReplaceMode::All));
+    }
+    VERIFY(!path.is_empty());
+    path = TRY(String::formatted("{}{}", devtmpfs_base_path, path));
+    mode_t old_mask = umask(0);
+    if (block_device)
+        TRY(Core::System::create_block_device(path.bytes_as_string_view(), match.create_mode, major_number.value(), minor_number.value()));
+    else
+        TRY(Core::System::create_char_device(path.bytes_as_string_view(), match.create_mode, major_number.value(), minor_number.value()));
+    umask(old_mask);
+    TRY(prepare_permissions_after_populating_devtmpfs(path.bytes_as_string_view(), match));
+
+    auto result = TRY(device_node_family.registered_nodes().try_set(RegisteredDeviceNode { move(path), minor_number }, AK::HashSetExistingEntryBehavior::Keep));
+    VERIFY(result != HashSetResult::ReplacedExistingEntry);
+    if (result == HashSetResult::KeptExistingEntry) {
+        // FIXME: Is this an actual bug?
+        return Error::from_errno(EEXIST);
+    }
+    allocation_map.set(allocated_suffix_index, true);
+    return {};
+}
+
+ErrorOr<void> DeviceEventLoop::unregister_device(bool block_device, MajorNumber major_number, MinorNumber minor_number)
+{
+    if (!device_node_family_to_match_type(block_device, major_number, minor_number))
+        return {};
+    auto* family = find_device_node_family(block_device, major_number, minor_number);
+    if (!family) {
+        // FIXME: Is this an actual bug?
+        return Error::from_errno(ENODEV);
```

Comment:
Hmm, what if we try to remove a device node that wasn't even registered?

A scenario where this could happen is if we run the `DeviceMapper` program after the user generated a device node and after the hotplug event was discarded because of a previous instance that was either killed or crashed...

Reply:
Restarts is a valid scenario, but I feel like in that case we should rather be able to recover most of our internal device library manually. But I guess it’s fine for now.